service/history/ndc/history_replicator_test.go

```diff
@@ -357,3 +360,196 @@ func TestNewHistoryReplicator_newMutableState(t *testing.T) {
 	)
 	assert.NotNil(t, testReplicatorImpl.newMutableState(mockDomainCacheEntry, log.NewNoop()))
 }
+
+func TestApplyEvents_newReplicationTask_validateReplicateEventsRequest(t *testing.T) {
```

Comment:
It looks like all the errors are from `validateReplicateEventsRequest`. 
Might consider creating some function parameters and overwriting them in unit tests to mock these functions.
And you can write tests for individual functions separately to test those functions referenced by ApplyEvents.
Here is an example you can follow: https://github.com/uber/cadence/pull/5971

Reply:
Amazing idea! I've made refactors according to your advice. Please take another look when you got some time. 🙏